### PR TITLE
Configurable exception handling

### PR DESCRIPTION
**Pull request by abhikuhikar**
Added ThunkOptions to configure exception handler which can be used instead of default handler which throws EvaluationException.
Added a class ThunkFactory which takes ThunkOptions as an argument. All the Thunk functions are moved to this class.
Added ThunkOptions property to CompileOptions.
Added a unit test to test the custom exception handler.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: lang/src/org/partiql/lang/eval/CompileOptions.kt [hunk 1]
```diff
@@ -95,7 +97,7 @@ data class CompileOptions private constructor (
         fun undefinedVariable(value: UndefinedVariableBehavior) = set { copy(undefinedVariable = value) }
         fun projectionIteration(value: ProjectionIterationBehavior) = set { copy(projectionIteration = value) }
         fun rewriterMode(value: RewritingMode) = set { copy(rewritingMode = value) }
-
+        fun thunkOptions(value: (Throwable, SourceLocationMeta?) -> Nothing) = set { copy(thunkOptions = ThunkOptions(value))}
```
Comment (dlurton) [inline lang/src/org/partiql/lang/eval/CompileOptions.kt:100]: In looking at this now I think we need to introduce a [`typealias`](https://kotlinlang.org/docs/reference/type-aliases.html):

```Kotlin
typealias ThunkExceptionHandler = (Throwable, SourceLocationMeta?) -> Nothing)
```

Then use the name `ThunkExceptionHandler` any place `(Throwable, SourceLocationMeta?) -> Nothing)` is used.

Also, I think the type of the `value` here should be `ThunkOptions`.

Furthermore, for consistency, I think `ThunkOptions` require builders following the pattern as `CompileOptions` and `EvaluationSession`.

File: lang/src/org/partiql/lang/eval/Thunk.kt [hunk 2]
```diff
@@ -37,167 +38,186 @@ typealias ThunkEnv = (Environment) -> ExprValue
...
- * Creates a [ThunkEnv], which handles exceptions by wrapping them into an [EvaluationException] which uses
- * [handleException] to handle exceptions appropriately.
+ * A thunk excepting a configurable exception handler (([Throwable], [SourceLocationMeta]?) -> [Nothing])
+ *
+ * The default exception handler wraps any [Throwable] exception and throws [EvaluationException]
  *
- * Literal lambdas passed to this function as [t] are inlined into the body of the function being returned, which
- * reduces the need to create additional call contexts.  The lambdas passed as [t] may not contain non-local returns
- * (`crossinline`).
+ * This name was chosen because it is a thunk which accepts exception handler options as its arguments.
```
Comment (dlurton) [inline lang/src/org/partiql/lang/eval/Thunk.kt:45]: I don't think this line is needed.

File: lang/src/org/partiql/lang/eval/Thunk.kt [hunk 3]
```diff
@@ -37,167 +38,186 @@ typealias ThunkEnv = (Environment) -> ExprValue
 typealias ThunkEnvValue<T> = (Environment, T) -> ExprValue
 
 /**
- * Creates a [ThunkEnv], which handles exceptions by wrapping them into an [EvaluationException] which uses
- * [handleException] to handle exceptions appropriately.
+ * A thunk excepting a configurable exception handler (([Throwable], [SourceLocationMeta]?) -> [Nothing])
```
Comment (dlurton) [inline lang/src/org/partiql/lang/eval/Thunk.kt:41]: Perhaps reword this to simply:  "Options for thunk construction."

File: lang/src/org/partiql/lang/eval/Thunk.kt [hunk 4]
```diff
@@ -37,167 +38,186 @@ typealias ThunkEnv = (Environment) -> ExprValue
...
-    return { env ->
-        handleException(sourceLocationMeta) {
-            t(env)
-        }
-    }
+class ThunkOptions(
+        val handleException: (Throwable, SourceLocationMeta?) -> Nothing = defaultExceptionHandler // <-- the default has the same behavior we currently have.
+)
+
+val defaultExceptionHandler: (Throwable, SourceLocationMeta?) -> Nothing = { e, sourceLocation ->
```
Comment (dlurton) [inline lang/src/org/partiql/lang/eval/Thunk.kt:51]: Make `internal`.

File: lang/src/org/partiql/lang/eval/Thunk.kt [hunk 5]
```diff
@@ -37,167 +38,186 @@ typealias ThunkEnv = (Environment) -> ExprValue
...
+    throw EvaluationException(
+        "Internal error, $message",
+        errorContext = errorContextFrom(sourceLocation),
+        cause = e,
+        internal = true)
 }
 
 /**
- * Similar to [thunkEnv], but creates a [ThunkEnvValue] instead.
+ * Contains all the thunk methods.
```
Comment (dlurton) [inline lang/src/org/partiql/lang/eval/Thunk.kt:61]: perhaps: "Provides methods for constructing new thunks according to the specified [ThunkOptions]."?

File: lang/src/org/partiql/lang/eval/Thunk.kt [hunk 6]
```diff
@@ -37,167 +38,186 @@ typealias ThunkEnv = (Environment) -> ExprValue
...
+ *
+ * Accepts [ThunkOptions] as an argument.
```
Comment (dlurton) [inline lang/src/org/partiql/lang/eval/Thunk.kt:63]: Not needed.

**Comment by dlurton** [comment]
To fix the failing travis build, merge from `master`.

File: lang/src/org/partiql/lang/eval/CompileOptions.kt (hunk 1 above)
Comment (abhikuhikar) [reply lang/src/org/partiql/lang/eval/CompileOptions.kt:100]: Okay. I will incorporate these.

Comment (abhikuhikar) [reply lang/src/org/partiql/lang/eval/CompileOptions.kt:100]: I have made these changes.

File: lang/src/org/partiql/lang/eval/Thunk.kt [hunk 7]
```diff
@@ -37,167 +38,221 @@ typealias ThunkEnv = (Environment) -> ExprValue
 typealias ThunkEnvValue<T> = (Environment, T) -> ExprValue
 
 /**
- * Creates a [ThunkEnv], which handles exceptions by wrapping them into an [EvaluationException] which uses
- * [handleException] to handle exceptions appropriately.
+ * A type alias for an exception handler
```
Comment (dlurton) [inline lang/src/org/partiql/lang/eval/Thunk.kt:41]: Needs a `.`

File: lang/src/org/partiql/lang/eval/Thunk.kt [hunk 8]
```diff
@@ -37,167 +38,221 @@ typealias ThunkEnv = (Environment) -> ExprValue
...
  *
- * Literal lambdas passed to this function as [t] are inlined into the body of the function being returned, which
- * reduces the need to create additional call contexts.  The lambdas passed as [t] may not contain non-local returns
- * (`crossinline`).
+ * The default exception handler wraps any [Throwable] exception and throws [EvaluationException]
  */
-internal inline fun thunkEnv(metas: MetaContainer, crossinline t: ThunkEnv): ThunkEnv {
-    val sourceLocationMeta = metas.find(SourceLocationMeta.TAG) as? SourceLocationMeta
+data class ThunkOptions private constructor(
+        val handleException: ThunkExceptionHandler = defaultExceptionHandler // <-- the default has the same behavior we currently have.
```
Comment (dlurton) [inline lang/src/org/partiql/lang/eval/Thunk.kt:51]: I don't think this comment is needed.

File: lang/src/org/partiql/lang/eval/CompileOptions.kt [hunk 9]
```diff
@@ -63,7 +64,8 @@ enum class RewritingMode {
 data class CompileOptions private constructor (
         val undefinedVariable: UndefinedVariableBehavior,
         val projectionIteration: ProjectionIterationBehavior = ProjectionIterationBehavior.FILTER_MISSING,
-        val rewritingMode: RewritingMode = RewritingMode.DEFAULT
+        val rewritingMode: RewritingMode = RewritingMode.DEFAULT,
+        val thunkOptions: ThunkOptions = ThunkOptions.build{}
```
Comment (dlurton) [inline lang/src/org/partiql/lang/eval/CompileOptions.kt:68]: `ThunkOptions.build{}` works but perhaps `ThunkOptions.standard()` is a little better.
